corehq/util/download.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from wsgiref.util import FileWrapper
+
+from django.http import StreamingHttpResponse, HttpResponse
+from werkzeug.http import parse_range_header
+
+from corehq.util.files import safe_filename_header
+
+
+class RangedFileWrapper:
+    """
+    Wraps a file like object with an iterator that runs over part (or all) of
+    the file defined by start and stop. Blocks of block_size will be returned
+    from the starting position, up to, but not including the stop point.
+
+    Based off: https://github.com/satchamo/django/commit/2ce75c5c4bee2a858c0214d136bfcd351fcde11d
+    """
+    block_size = 8192
+
+    def __init__(self, filelike, start=0, stop=float("inf"), block_size=None):
+        self.filelike = filelike
+        self.block_size = block_size or RangedFileWrapper.block_size
+        self.start = start
+        self.stop = stop
+        if hasattr(filelike, 'close'):
+            self.close = filelike.close
+
+    def __iter__(self):
+        if hasattr(self.filelike, 'seek'):
+            self.filelike.seek(self.start)
+        else:
+            pos_tmp = 0
+            while pos_tmp < self.start:
+                data = self.filelike.read(min(self.block_size, self.start - pos_tmp))
+                pos_tmp += len(data)
+                if not data or pos_tmp == self.start:
+                    break
+
+        position = self.start
+        while position < self.stop:
+            data = self.filelike.read(min(self.block_size, self.stop - position))
+            if not data:
+                break
+
+            yield data
+            position += self.block_size
+
+
+def get_download_response(payload, content_length, content_format, filename, request=None):
+    """
+    :param payload: File like object.
+    :param content_length: Size of payload in bytes
+    :param content_format: ``couchexport.models.Format`` instance
+    :param filename: Name of the download
+    :param request: The request. Used to determine if a range response should be given.
+    :return: HTTP response
+    """
+    ranges = None
+    if "HTTP_RANGE" in request.META:
```

Comment:
I believe this will error if no `request` is passed in and the default value of `request=None` is used